pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -747,3 +772,22 @@ func addressSubset(addresses []v1.EndpointAddress, maxNum int) []v1.EndpointAddr
 	}
 	return addresses[0:maxNum]
 }
+
+// managedByController returns true if the endpoint is managed by this endpoints controller
+func managedByController(endpoint *v1.Endpoints) bool {
+	if _, ok := endpoint.Annotations[resourcelock.LeaderElectionRecordAnnotationKey]; ok {
+		// when there are multiple controller-manager instances,
+		// we observe that it will delete leader-election endpoints after 5min
+		// and cause re-election
+		// so skip the delete here
+		// as leader-election only have endpoints without service
+		return false
+	}
+	// endpoints without managed fields are not handled by this controller
+	// for backwards compatibility
+	if len(endpoint.ManagedFields) == 0 {
+		return false
+	}
+	// manage all endpoints created by this controller
+	return endpoint.ManagedFields[0].Manager == controllerName
```

Comment:
@apelisse is this the correct way to check that this Endpoint object was created by this controller?
 is the ManagedFields array sorted and entry 0 is the first one?

Reply:
No, I don't know if you can make that assumption.

Here's the code that sorts the entries: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go#L194-L224